java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -53,97 +60,87 @@ import arcs.core.util.Time
-     * @property idGenerator used to generate unique IDs for newly stored entities.
-     */
-    suspend fun <T : Entity> createCollectionHandle(
-        mode: HandleMode,
-        baseName: String,
-        entitySpec: EntitySpec<T>,
-        storageKey: StorageKey,
-        ttl: Ttl = Ttl.Infinite,
-        idGenerator: Id.Generator = Id.Generator.newSession()
-    ): Handle =
-        idGenerator.handleName(baseName).let { name ->
-            CollectionHandle(
-                name = idGenerator.handleName(baseName),
-                entitySpec = entitySpec,
-                storageProxy = collectionStoreProxy(storageKey, entitySpec),
-                entityPreparer = EntityPreparer(name, idGenerator, entitySpec.SCHEMA, ttl, time),
-                dereferencerFactory = dereferencerFactory
-            )
-        }.let {
-            when (mode) {
-                HandleMode.Read -> when (entitySpec.SCHEMA.query) {
-                    null -> object : ReadCollectionHandle<T> by it {}
-                    else -> object : ReadQueryCollectionHandle<T, Any> by it {}
+        return when (spec.containerType) {
+            HandleContainerType.Singleton -> {
+                val singletonHandle = SingletonHandle(
+                    name = handleName,
+                    spec = spec,
+                    storageProxy = singletonStoreProxy(storageKey, spec.entitySpec)
+                        as SingletonProxy<out Referencable>,
+                    storageAdapter = storageAdapter,
+                    dereferencerFactory = dereferencerFactory
+                )
+                when (spec.mode) {
+                    HandleMode.Read -> object : ReadSingletonHandle<T> by singletonHandle {}
+                    HandleMode.Write -> object : WriteSingletonHandle<T> by singletonHandle {}
+                    HandleMode.ReadWrite ->
+                        object : ReadWriteSingletonHandle<T> by singletonHandle {}
                 }
-                HandleMode.Write -> object : WriteCollectionHandle<T> by it {}
-                HandleMode.ReadWrite -> when (entitySpec.SCHEMA.query) {
-                    null -> object : ReadWriteCollectionHandle<T> by it {}
-                    else -> object : ReadWriteQueryCollectionHandle<T, Any> by it {}
+            }
+            HandleContainerType.Collection -> {
+                val collectionHandle = CollectionHandle(
+                    name = handleName,
+                    spec = spec,
+                    storageProxy = collectionStoreProxy(storageKey, spec.entitySpec)
+                        as CollectionProxy<out Referencable>,
+                    storageAdapter = storageAdapter,
+                    dereferencerFactory = dereferencerFactory
+                )
+                when (spec.mode) {
+                    HandleMode.Read -> when (spec.entitySpec.SCHEMA.query) {
+                        null -> object : ReadCollectionHandle<T> by collectionHandle {}
```

Comment:
why is the mode allowed to be null? Shouldn't we throw an error instead?

Reply:
Oops, never mind, I didn't see the nested when, I shouldn't review things from my iphone. :(